File: practices/chapter9_sort/python/ch_9_cocktail_sort.py
def cocktail_sort(data_list):
  res = data_list
  length = len(data_list)
  start_index = 0
  end_index = length - 1
  is_sorted = True

  while is_sorted:
    is_sorted = False
    # 往右比較與對調
    for i in range(end_index):
      if res[i] > res[i + 1]:
        res[i], res[i + 1] = res[i + 1], res[i]
        is_sorted = True
      pass
    print('To right sorted: ', res)

    if not is_sorted: # 如果不用對調
      break

    # 往左比較與對調
    for i in range(end_index - 1, start_index - 1, -1):
      if res[i] > res[i + 1]:
        res[i], res[i + 1] = res[i + 1], res[i]
        is_sorted = True
      pass
    print('To left sorted: ', res)
    start_index = start_index + 1

  return res

def main():
  data_list = [1, 3, 2, 8, 7]
  print('Origin data', data_list)
  print('Cocktail sorted: ', cocktail_sort(data_list))

if __name__ == "__main__":
    main()

File: practices/chapter9_sort/js/ch_9_cock_tail_sort.js
function cocktailSort(dataList=[]) {
  res = [...dataList]
  length = dataList.length
  startIdx = 0
  endIdx = length - 1
  isSorted = true

  while(isSorted) {
    isSorted = false

    for (let i = 0; i <= endIdx; i++) {
      // 往右查
      if(res[i] > res[i + 1]) {
        [res[i], res[i + 1]] = [res[i + 1], res[i]]
        isSorted = true
      }
    }
    // 這一輪的往右對調完了(最右已經是最大的)，將end index往左
    endIdx --

    if(!isSorted) {
      break
    }

    for (let i = endIdx; i >= startIdx; i--) {
      // 往左查
      if(res[i + 1] < res[i]) {
        [res[i], res[i + 1]] = [res[i + 1], res[i]]
        isSorted = true
      }
    }
    // 這一輪的往左對調完了(最左已經是最小的)，將start index往右
    startIdx ++
  }

  return res
}

function main() {
  const dataList = [9, 3, 2, 8, 7, 10, 6, 5, 4, 1, ]
  console.log('Origin data: ', dataList)
  console.log('Sorted: ', cocktailSort(dataList))
}

main()